text/1154-public-api-for-render-tree-based-scope-exploration.md

```diff
@@ -0,0 +1,435 @@
+  let scope = getScope();
+
+  // search up until we find our key
+  for (let i = 0; i < scope.entries.length; i++) {
+    let entry = scope.entries[i];
+
+    if (!Array.isArray(entry)) continue;
+    if (entry[0] === key) return entry[1];
+  }
+}
+
+export class Consume extends Component {
+  get context() {
+    assert(`Must pass a key`, this.args.key);
+
+    return consume(this.args.key);
+  }
+
+  <template>
+    {{yield this.context}}
+  </template>
+}
+```
+
+This supports both class providing and consuming as well as template-based providing and consuming.
+
+```gjs
+import { Consume, Provide, consume, provide } from 'hypothetical ember-prototype-context library';
+
+let id = 0;
+class StateA { foo = 'a'; id = id++; }
+class StateB { foo = 'b' }
+
+const newA = () => new StateA();
+const newB = () => new StateB();
+
+class CustomProvide extends Component {
+  @provide a = new StateA();
+}
+
+const CustomConsume = <template>
+  {{yield (consume @key)}}
+</template>;
+
+<template>
+  <Provide @key={{StateA}} @context={{ (newA) }}>
+    <Consume @key={{StateA}} as |a|>
+      {{a.foo}} === "a"
+      {{a.id}} === 0
+    </Consume>
+
+    {{#let (consume StateA) as |a|}}
+      {{a.foo}} === "a"
+      {{a.id}} === 0;
+    {{/let}}
+  </Provide>
+
+  <Provide @key={{StateA}} @context={{ (newA) }}>
+    {{#let (provide StateB (newB)) as |b|}}
+      {{b.foo}} === "b"
+
+      <Consume @key={{StateB}} as |b|>
+        {{b.foo}} === "b"
+
+        <Consume @key={{StateA}} as |a|>
+          {{a.foo}} === "a"
+          {{a.id}} === 1
+        </Consume>
+      </Consume>
+    {{/let}}
+
+    <Consume @key={{StateA}} as |a|>
+      {{a.foo}} === "a"
+      {{a.id}} === 1
+    </Consume>
+
+    <CustomConsume @key={{StateA}} as |a|>
+      {{a.foo}} === "a"
+      {{a.id}} === 1;
+    </CustomConsume>
+  </Provide>
+
+
+  <CustomProvide>
+    {{#let (consume StateA) as |a|}}
+      {{a.foo}} === "a"
+      {{a.id}} === 2;
+    {{/let}}
+  </CustomProvide>
+
+</template>
+```
+
+
+## Drawbacks
+
+We already track a "stack" in our current renderer, but depending on how this is implemented, it could accidentally double the memory usage of that stack -- however most "scopes" will be empty / reused between render nodes, as what is in the scope is not expected to change that often.
+
+Depending on implementation, there _could_ be a userland-induced memory leak, so we'll want to explore how to ensure the scope entries don't endlessly grow over time (we already manage this with the current renderer's stack (see second alternative below)).
+
+Adding more APIs is always more for people to learn, but many frontend frameworks already have similar features, so developers should have an easy time picking this up if they wish (or the future features enabled by this API).
+
+## Alternatives
+
+- [ember-provide-consume-context](https://github.com/customerio/ember-provide-consume-context)
```

Comment:
I don't see `ember-provide-consume-context` as an alternative to scope. It *is* a different way of tackling the context RFC by not using managers.

This could be rephrased to 

```suggestion
Instead of making the low level reactivity public API, we could adopt a public API for context and keep the internals private. This is the approach of the common addon [ember-provide-consume-context](https://github.com/customerio/ember-provide-consume-context) which relies on modifying opcodes in Ember's VM and keeps separate non-observable stack of context.
```